src/a.tex

```diff
@@ -229,19 +229,27 @@ \section{Load-Reserved/Store-Conditional Instructions}
 
 An SC instruction can never be observed by another RISC-V hart
 before the LR instruction that established the reservation.
+
+\begin{commentary}
 The LR/SC
 sequence can be given acquire semantics by setting the {\em aq} bit on
 the LR instruction.  The LR/SC sequence can be given release semantics
-by setting the {\em rl} bit on the SC instruction.  Setting the {\em
-  aq} bit on the LR instruction, and setting both the {\em aq} and the {\em
-  rl} bit on the SC instruction makes the LR/SC sequence sequentially
-consistent, meaning that it cannot be reordered with earlier or
-later memory operations from the same hart.
-
-If neither bit is set on both LR and SC, the LR/SC sequence can be
+by setting the {\em rl} bit on the SC instruction.  Assuming
+suitable mappings for other atomic operations, setting the
+{\em aq} bit on the LR instruction, and setting the
+{\em rl} bit on the SC instruction makes the LR/SC
+sequence sequentially consistent in the C++ {\em memory\_order\_seq\_cst}
+sense. Such a sequence does not act as a fence for ordering ordinary
+load and store instructions before and after the sequence. Specific
```

Comment:
I don't think this is true, since the PPO rules order the acquire with all younger instructions and order the release with all older instructions.

In any case, this is why I'd like to constrain the textual changes to the mapping section in the appendix, rather than modifying the normative text in the A chapter that defines the instructions.  @daniellustig can you spare some time to work with Hans to organize this PR in a way we'd all find agreeable?

Reply:
I may have misunderstood your request. I did enclose this in a commentary block, which presumably is non-normative?

Regarding the technical issue: If I have

s; lr.aq; sc.rl; l

where the lr and sc refer to the same location, but the s and l refer to different locations, I don't believe there is anything to prevent s and the l from being reordered in between the lr and sc, and then past each other. In that sense, the lr;sc does not act as a fence. You get the same phenomenon with programming-language-level locks and relaxed atomic operations (in spite of some amount of code the incorrectly assumes that critical sections are fences)

If you know of a better way to state this, I would be happy to do so. I think this is worth stating in commentary because many people find it surprising, and I think the Linux kernel has tried to strengthen some atomics primitives to avoid it, something I would argue against.

Reply:
Sorry, @hboehm, I misread the TeX: you did, indeed, enclose this note within the commentary, thus making it non-normative.

On the technical matter, I believe the RVWMO PPO rules contradict your position.  Rule 6 says `s` and `lr.aq` are ordered before `sc.rl`.  Rule 5 says `l` and `sc.rl` are ordered after `lr.aq`.  So `s` and `l` are ordered.

Reply:
As to the technical matter, I agree on further reflection that I can't find a PPO rule that suggests `s` and `l` are ordered.